src/hash_table_internal.h

```diff
@@ -46,32 +43,67 @@
  */
 struct ly_ht_rec {
     uint32_t hash;        /* hash of the value */
-    int32_t hits;         /* collision/overflow value count - 1 (a filled entry has 1 hit,
-                           * special value -1 means a deleted record) */
-    unsigned char val[1]; /* arbitrary-size value */
-} _PACKED;
+    uint32_t next;        /* index of next collision */
+    unsigned char val[]; /* arbitrary-size value */
+};
+
+struct ly_ht_bucket {
+    uint32_t first;
+    uint32_t last;
+};
 
 /**
  * @brief (Very) generic hash table.
  *
- * Hash table with open addressing collision resolution and
- * linear probing of interval 1 (next free record is used).
- * Removal is lazy (removed records are only marked), but
- * if possible, they are fully emptied.
+ * The hash table is composed of a table of indexes that references the
+ * first record. The records contain a next index that references the
+ * next record in case of collision.
+ * The free records are chained starting from first_free_rec.
```

Comment:
The original description was meant to formally characterize the hash table implementation so that it can be understood and changed easier. It may be my fault, but I understand almost nothing from this, seems very generic. It is fine if you do not use any common algorithm names but then I would expect a brief description of the functionality instead.

Reply:
You are right, I'll try to enhance the description.